text/0025-configuration-policy.md

```diff
@@ -0,0 +1,197 @@
+overridding the `onConstructCreation` method. `onConstructCreation` is called
+for every `Construct` added to the `Stack`, and receives the same properties
+used to create the `Construct`. These properties can be altered or validated to
+enforce default behaviors.
+
+The following demonstrates a Stack that enforces encryption for S3 buckets by
+default.
+
+```ts
+// Base-line Stack to act as a subclass for all other Stacks.
+export class FooBarCompanyStack extends Stack {
+  public onConstructCreation(className: string, id: string, props: any): any {
+    if (className == 'aws-s3.Bucket') {
+      const bucketProps = <BucketProps>props;
+      if (bucketProps.encryption === undefined || bucketProps.encryption === BucketEncryption.UNENCRYPTED) {
+        return {
+          encryption: BucketEncryption.S3_MANAGED,
+          ...props,
+        };
+      }
+    }
+    return props;
+  }
+}
+// App-specific Stack which receives the defaults.
+export class MyAppStack extends FooBarCompanyStack {
+    constructor(scope: Construct, id: string, props?: StackProps) {
+        super(scope, id, props);
+        new Bucket(this, 'myBucket'); // Will automatically get S3-managed encryption.
+  }
+}
+```
+
+In addition to modifying the properties of the `Construct` being created, the
+`onConstructCreation` hook can be used to create entirely new resources. This
+example enforces a convention of each SQS Queue having its own dead-letter
+queue.
+
+```ts
+export class FooBarCompanyStack extends Stack {
+  public onConstructCreation(className: string, id: string, props: any): any {
+    if (className === 'aws-sqs.Queue') {
+      const queueProps = <QueueProps>props;
+      if (!id.includes("DLQ") && queueProps.deadLetterQueue === undefined) {
+        const dlq = new Queue(this, id + "DLQ");
+        return {
+          deadLetterQueue: { queue: dlq, maxReceiveCount: 3 },
+          ...props,
+        };
+      }
+    }
+    return props;
+  }
+}
+```
+
+# Motivation
+
+Teams, organizations, and companies may have need to enforce or apply default
+best practices to CDK-managed infrastructure. This approach will provide a
+consistent mechanism to apply these practices, while keeping the actual
+business-logic-containing Stacks lightweight.
+
+See the discussion in <https://github.com/aws/aws-cdk/issues/3235> for more
+context.
+
+# Basic Example
+
+_See README section._
+
+# Design Summary
+
+The approach adds a new hook (`onConstructCreation`) to the `Stack` class,
+which will be called by every L2 on construction. This hook receives the same
+inputs as the Construct's constructor, and enables users to adjust or override
+any of the input props, as well as throw errors or create ancillary resources.
+
+# Detailed Design
+
+A method is added to the `Stack` class with the following signature and default
+implementation:
+
+```ts
+public onConstructCreation(className: string, id: string, props: any): any {
```

Comment:
- The term "Construct" is too broad (and BTW we use it as all lowercase like "class"). This hook is just for "resources", not any construct, so I would use this semantic.
- We use `onXxx` across L2s for CloudWatch events, so I am not sure this prefix is a good one, but on the other hand... it is a lifecycle hook.
- "Creation" feels convoluted. Can we just do "Create"?

So: `onResourceCreate` perhaps?